sympy/combinatorics/fp_groups.py

```diff
@@ -1003,4 +1018,389 @@ def first_in_class(C, Y=[]):
+        j -= 1
+    if j < i:
+        modified_coincidence(C, f, b, f_p**-1*b_p)
+    elif j == i:
+        # deduction making
+        table[f][C.A_dict[word[i]]] = b; table[b][C.A_dict_inv[word[i]]] = f
+        C.P[f][C.A_dict[word[i]]] = f_p**-1*b_p
+        C.P[b][C.A_dict_inv[word[i]]] = b_p**-1*f_p
+    # otherwise scan is incomplete and yields no information
+
+def modified_rep(C, k):
+    p = C.p
+    p_p = C.p_p
+    lamda = k
+    rho = p[lamda]
+    # `s` is used to trace back the compression path
+    s = [None]*len(p)
+    while rho != lamda:
+        s[rho] = lamda
+        lamda = rho
+        rho = p[lamda]
+    rho = s[lamda]
+    while rho != k:
+        mu = rho
+        rho = s[mu]
+        p[rho] = lamda
+        p_p[rho] = p_p[rho]*p_p[mu]
+    return lamda
+
+def modified_merge(C, k, lamda, w, q):
+    p = C.p
+    p_p = C.p_p
+    phi = C.rep(k)
+    psi = C.rep(lamda)
+    # can't use `max/min` function as in `CosetTable.merge`
+    if phi > psi:
+        p[phi] = psi
+        p_p[phi] = p_p[k]**-1*w*p_p[lamda]
+        q.append(psi)
+    elif psi > phi:
+        p[psi] = phi
+        p_p[psi] = p_p[lamda]**-1*w**-1*p_p[k]
+        q.append(psi)
+
+def modified_coincidence(C, alpha, beta, w):
+    A_dict = C.A_dict
+    A_dict_inv = C.A_dict_inv
+    table = C.table
+    p_p = C.p_p
+    l = 0
+    q = []
+    modified_merge(C, alpha, beta, w, q)
+    while len(q) > 0:
+        gamma = q.pop(0)
+        for x in A_dict:
+            delta = table[gamma][A_dict[x]]
+            if delta is not None:
+                table[delta][A_dict_inv[x]] = None
+                mu = C.rep(gamma)
+                nu = C.rep(delta)
+                if table[mu][A_dict[x]] is not None:
+                    v = p_p[delta]**-1*C.P[gamma][A_dict[x]]**-1*p_p[gamma]*C.P[mu][A_dict[x]]
+                    modified_merge(C, nu, table[mu][A_dict[x]], v, q)
+                elif table[nu][A_dict_inv[x]] is not None:
+                    v = p_p[gamma]**-1*C.P[gamma][A_dict[x]]*p_p[delta]*C.P[mu][A_dict_inv[x]]
+                else:
+                    table[mu][A_dict[x]] = nu; table[nu][A_dict_inv[x]] = mu
+                    v = p_p[gamma]**-1*C.P[gamma][A_dict[x]]*p_p[delta]
+                    C.P[mu][A_dict[x]] = v
+                    C.P[nu][A_dict_inv[x]] = v**-1
+
+
+# Pg 350, section 2.5.1 from [2]
+def elimination_technique_1(C):
+    rels = list(C.reidemeister_relators)
+    # the shorter relators are examined first so that generators selected for
+    # elimination will have shorter strings as equivalent
+    rels.sort(reverse=True)
+    gens = list(C.schreier_generators)
+    redundant_gens = {}
+    contained_gens = []
+    # examine each relator in relator list for any generator occuring exactly
+    # once
+    next_i = False
+    for i in range(len(rels) -1, -1, -1):
+        rel = rels[i]
+        for gen in redundant_gens:
+            if any([gen.array_form[0][0] == r[0] for r in rel.array_form]):
```

Comment:
This is a loop. It would be better to assign `gen.array_form[0][0]` to a local variable.